include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -234,74 +236,141 @@ namespace VectorTools
+                    shifted_view, inverse_jacobians[i], old_value);
+
+                  for (unsigned int j = 0; j < dim; ++j)
+                    shifted_view[j] *= jacobians[i].determinant();
+                }
+              break;
 
-          case FiniteElementData<dim>::Hdiv:
-            // See Monk, Finite Element Methods for Maxwell's Equations,
-            // p. 79ff, formula (3.77) and Lemma 3.59.
-            // For given mapping F_K: \hat K \to K, we have to transform
-            //  \hat w = det(dF_K) (dF_K)^{-1} w\circ F_K
+            case FiniteElementData<dim>::H1:
+              DEAL_II_FALLTHROUGH;
+            case FiniteElementData<dim>::L2:
+              // See Monk, Finite Element Methods for Maxwell's Equations,
+              // p. 77ff, formula (3.74).
+              // For given mapping F_K: \hat K \to K, we have to transform
+              //  \hat p = p\circ F_K
+              //  i.e., do nothing.
+              break;
 
-            fe_values_jacobians.reinit(cell);
-            for (unsigned int i = 0; i < function_values.size(); ++i)
-              {
-                const auto &jacobians =
-                  fe_values_jacobians.get_present_fe_values().get_jacobians();
-                const auto &inverse_jacobians =
-                  fe_values_jacobians.get_present_fe_values().get_inverse_jacobians();
-
-                // value[m] <- sum inverse_jacobians[m][n] * old_value[n]:
-                const auto old_value = function_values[i];
-                TensorAccessors::contract<1, 2, 1, dim>(
-                  function_values[i], inverse_jacobians[i], old_value);
-
-                for (unsigned int j = 0; j < n_components; ++j)
-                  function_values[i][j] *= jacobians[i].determinant();
-              }
-            break;
+            default:
+              // In case we deal with an unknown conformity, just assume we
+              // deal with a Lagrange element and do nothing.
+              break;
 
-          case FiniteElementData<dim>::H1:
-            DEAL_II_FALLTHROUGH;
-          case FiniteElementData<dim>::L2:
-            // See Monk, Finite Element Methods for Maxwell's Equations,
-            // p. 77ff, formula (3.74).
-            // For given mapping F_K: \hat K \to K, we have to transform
-            //  \hat p = p\circ F_K
-            //  i.e., do nothing.
-            //
-            break;
+            } /*switch*/
+        }; /* lambda function transform */
 
-          default:
-            // In case we deal with an unknown conformity, just assume we
-            // deal with a Lagrange element and do nothing.
-            break;
+        // Before we can average, we have to transform all function values
+        // from the real cell back to the unit cell. We query the finite
+        // element for the correct transformation. Matters get a bit more
+        // complicated because we have to apply said transformation for
+        // every base element.
```

Comment:
I just wanted to share this beautiful C++14 gem to get recursive lambdas to work. This obviously has to be refactored.

@tjhei @bangerth Do you see some potential use for such a generic traversal over (base) finite elements? If yes, we might consider including this primitive in `FiniteElement` (`FiniteElement::for_each_base_element(<lambda>)` or similar).